### PR TITLE
All nwaku nodes use own account

### PR DESCRIPTION
**Pull request by stubbsta**
This PR implements the use of a separate testnet account for each nwaku node.
The nodes require this to be able to generate the RLN keystore without having collisions (https://github.com/waku-org/waku-simulator/issues/52).
The following changes have been made:
- generates testnet accounts equal to the number of nwaku node services
- writes the anvil config (which includes the generated private keys) to a shared volume
- nwaku nodes retrieve a private key form the shared volume based on their own index

### CODE REVIEW

**Comment by stubbsta** [comment]
The current implementation only generates the required number of accounts and writes those private keys to a file, this is more dynamic but it does require some overhead to parse the anvil config json and write only the private keys to a file.
The private keys generated by anvil are deterministic so I could put them in a constant variable if this method is preferred.

File: getPrivateKeys.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
```
Comment (alrevuelta) [inline getPrivateKeys.sh:2]: Can we avoid having this extra script? Its another moving part to mount, track etc.

The content of this script can be in `run_nwaku` I guess?

File: docker-compose.yml [hunk 2]
```diff
@@ -21,11 +21,14 @@ services:
...
         --chain-id=1337
-        --accounts=1
+        --accounts=${NUM_NWAKU_NODES:-5}
         --allow-origin=*
         --block-time=12
         --silent
         --config-out=anvil-config.txt
+    volumes:
+      - privatekeys-volume:/shared
+      - ./getPrivateKeys.sh:/getPrivateKeys.sh
```
Comment (alrevuelta) [inline docker-compose.yml:31]: can we remove this script?

File: docker-compose.yml [hunk 3]
```diff
@@ -216,4 +219,7 @@ services:
       - redis
       - foundry
     networks:
-      - simulation
\ No newline at end of file
+      - simulation
+
+volumes:
```
Comment (alrevuelta) [inline docker-compose.yml:224]: is this needed?

File: getPrivateKeys.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Read the JSON file
+json_content=$(cat anvil-config.txt)
+
+# Extract private_keys array values using jq
+private_keys=$(echo "$json_content" | jq -r '.private_keys[]')
+
+# Write private keys to a new file for easier indexing
+echo "Writing private keys to file"
+echo "$private_keys" > /shared/private-keys.txt
```
Comment (alrevuelta) [inline getPrivateKeys.sh:15]: we can avoid also this private-keys files. anvil-config has everything we need, no need IMHO to have this intermediate file.

File: run_nwaku.sh [hunk 5]
```diff
@@ -33,29 +35,42 @@ else
     fi
 fi
 
+#Function to get the index of the container and use it to retrieve a private key to be used to generate the keystore
+get_private_key(){
```
Comment (alrevuelta) [inline run_nwaku.sh:39]: and here we can direactly have `getPrivateKeys.sh` content. where anvil-config is loaded, parsed and where INDEX is taken.

**Review by alrevuelta — CHANGES_REQUESTED** [review]
thanks! a minor comment to simplify things a bit.

File: docker-compose.yml (hunk 3 above)
Comment (stubbsta) [reply docker-compose.yml:224]: The shared volume is required for different services to have access to read and write to the same data/object. It doesn't seem possible to have a variable shared between containers. It is the only way I could find for the nwaku containers to have access to the private keys generated by anvil. The alternative is to hardcode the private keys to a variable, like you suggested initially.

File: getPrivateKeys.sh (hunk 1 above)
Comment (stubbsta) [reply getPrivateKeys.sh:2]: TL;DR: let me see what I can do

I implemented it like this to try and simplify things because my initial implementation had everything in the run_nwaku script, but this was tricky and messy because the nwaku image runs on Alpine which can't work with arrays and indexing without installing bash, which then requires a you to execute a script to install bash and then execute the array type commands in a script which is called with the bash script and yeah, things just got messy, but let me try again and see if I can simplify it.

Comment (stubbsta) [reply getPrivateKeys.sh]: The script has been removed and all logic from the script has been moved to run_nwaku.sh
Because of Alpine's default /bin/ash shell we now need to iterate through the list of private keys, which isn't ideal, but it only needs to be done once at startup so it's not too bad.

File: getPrivateKeys.sh (hunk 4 above)
Comment (stubbsta) [reply getPrivateKeys.sh]: removed in latest commit https://github.com/waku-org/waku-simulator/pull/57/commits/71882e78b56af87b10bbba8d5f2cd3986689aff0

File: run_nwaku.sh (hunk 5 above)
Comment (stubbsta) [reply run_nwaku.sh:40]: Done with some modifications for /bin/ash script limitations (https://github.com/waku-org/waku-simulator/pull/57/commits/71882e78b56af87b10bbba8d5f2cd3986689aff0 )
